rust/pattern/variable/builder/thing_variable_builder.rs

```diff
@@ -36,52 +37,84 @@ pub trait ThingVariableBuilder {
         self,
         type_name: impl Into<String>,
         value: T,
-    ) -> Result<Variable, ErrorMessage>
+    ) -> Result<ThingVariable, ErrorMessage>
     where
         ErrorMessage: From<<T as TryInto<Value>>::Error>;
 
-    fn iid<T: TryInto<IIDConstraint>>(self, iid: T) -> Result<Variable, ErrorMessage>
+    fn iid<T: TryInto<IIDConstraint>>(self, iid: T) -> Result<ThingVariable, ErrorMessage>
     where
         ErrorMessage: From<<T as TryInto<IIDConstraint>>::Error>;
 
-    fn isa(self, isa: impl Into<IsaConstraint>) -> Result<Variable, ErrorMessage>;
+    fn isa(self, isa: impl Into<IsaConstraint>) -> Result<ThingVariable, ErrorMessage>;
 
-    fn eq(self, value: impl Into<Value>) -> Result<Variable, ErrorMessage>;
+    fn eq(self, value: impl Into<Value>) -> Result<ThingVariable, ErrorMessage>;
+    fn neq(self, value: impl Into<Value>) -> Result<ThingVariable, ErrorMessage>;
+    fn gt(self, value: impl Into<Value>) -> Result<ThingVariable, ErrorMessage>;
+    fn gte(self, value: impl Into<Value>) -> Result<ThingVariable, ErrorMessage>;
+    fn lt(self, value: impl Into<Value>) -> Result<ThingVariable, ErrorMessage>;
+    fn lte(self, value: impl Into<Value>) -> Result<ThingVariable, ErrorMessage>;
+    fn contains(self, value: impl Into<String>) -> Result<ThingVariable, ErrorMessage>;
+    fn like(self, value: impl Into<String>) -> Result<ThingVariable, ErrorMessage>;
 }
 
 impl<U: ThingConstrainable> ThingVariableBuilder for U {
     fn has<T: TryInto<Value>>(
         self,
         type_name: impl Into<String>,
         value: T,
-    ) -> Result<Variable, ErrorMessage>
+    ) -> Result<ThingVariable, ErrorMessage>
     where
         ErrorMessage: From<<T as TryInto<Value>>::Error>,
     {
-        Ok(self
-            .constrain_has(match value.try_into()? {
-                Value::Variable(variable) => HasConstraint::from((type_name.into(), *variable)),
-                value => HasConstraint::from((
-                    type_name.into(),
-                    ValueConstraint::new(Predicate::Eq, value),
-                )),
-            })
-            .into_variable())
-    }
-
-    fn iid<T: TryInto<IIDConstraint>>(self, iid: T) -> Result<Variable, ErrorMessage>
+        Ok(self.constrain_has(match value.try_into()? {
+            Value::Variable(variable) => HasConstraint::from((type_name.into(), *variable)),
+            value => {
+                HasConstraint::from((type_name.into(), ValueConstraint::new(Predicate::Eq, value)))
+            }
+        }))
+    }
+
+    fn iid<T: TryInto<IIDConstraint>>(self, iid: T) -> Result<ThingVariable, ErrorMessage>
     where
         ErrorMessage: From<<T as TryInto<IIDConstraint>>::Error>,
     {
-        Ok(self.constrain_iid(iid.try_into()?).into_variable())
+        Ok(self.constrain_iid(iid.try_into()?))
+    }
+
+    fn isa(self, isa: impl Into<IsaConstraint>) -> Result<ThingVariable, ErrorMessage> {
+        Ok(self.constrain_isa(isa.into()))
```

Comment:
That was because a macro expects `into_pattern`. It should really be a trait or, more likely, an `impl From<T> for Pattern`.

It should be clear here what type we convert to. `isa` is `impl Into<IsaConstraint>`, so `isa.into()` therefore returns `IsaConstraint`. Bare `into()` calls appear to be idiomatic Rust.